pkg/util/kubernetes/apiserver/events.go

```diff
@@ -7,103 +7,144 @@
 
 package apiserver
 
-//// Covered by test/integration/util/kube_apiserver/events_test.go
+//// Covered by test/integration/util/kube_apiserver/apiserver_test.go
 
 import (
 	"fmt"
 	"strconv"
 	"time"
 
-	"github.com/pkg/errors"
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/watch"
 
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-// LatestEvents retrieves all the cluster events happening after a given token.
-// First slice is the new events, second slice the modified events.
-// If the `since` parameter is empty, we query the apiserver's cache to avoid
-// overloading it.
-// https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#watch-list-289
-func (c *APIClient) LatestEvents(since string, eventReadTimeout time.Duration) ([]*v1.Event, []*v1.Event, string, error) {
-	var added, modified []*v1.Event
-
-	// If `since` is "" strconv.Atoi(*latestResVersion) below will panic as we evaluate the error.
-	// One could chose to use "" instead of 0 to not query the API Server cache.
-	// We decide to only rely on the cache as it avoids crawling everything from the API Server at once.
-	resVersionInt, err := strconv.Atoi(since)
-	if err != nil {
-		log.Errorf("The cached resourceVersion token %s, could not be parsed with: %s", since, err)
-		since = "0"
+// RunEventCollection will return the most recent events emitted by the apiserver.
+func (c *APIClient) RunEventCollection(srv string, st time.Time, eventReadTimeout int64, eventCardinalityLimit int64, resync int64, filter string) ([]*v1.Event, string, time.Time, error) {
+	var added []*v1.Event
+	syncTimeout := time.Duration(resync) * time.Second
+	// list if latestResVer is "" or if lastListTS is > syncTimeout
+	if srv == "" || time.Now().Sub(st) > syncTimeout {
+		evList, err := c.Cl.CoreV1().Events(metav1.NamespaceAll).List(metav1.ListOptions{
+			TimeoutSeconds:       &eventReadTimeout,
+			Limit:                eventCardinalityLimit,
+			IncludeUninitialized: false,
+			FieldSelector:        filter,
+		})
+		if err != nil {
+			log.Errorf("Error Listing events: %s", err.Error())
+			return nil, srv, st, err
+		}
+		for _, e := range evList.Items {
+			// List call returns a different type than the Watch call.
+			added = append(added, &e)
+		}
+		st = time.Now()
+		srv = evList.ResourceVersion
+		return added, srv, st, nil
 	}
-
-	log.Tracef("Starting watch of events with resourceVersion %s", since)
-
-	eventWatcher, err := c.Cl.CoreV1().Events(metav1.NamespaceAll).Watch(metav1.ListOptions{Watch: true, ResourceVersion: since})
+	// Start watcher with the most up to date RV
```

Comment:
just wondering how the RV interacts with `continue` if the `limit` of event is reached during the initial listing?

Reply:
We will get the `Limit` most recent events, and a continue token, that we do not use. So if we are down for some time and start processing events with an initial list that outgrows the Limit, some events won't be reported.